### PR TITLE
tls(Server) fix connectionListener and make alpnProtocol more compatible with node.js

### PR DESCRIPTION
**Pull request by cirospaciari**
### What does this PR do?

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [X] Code changes

### How did you verify your code works?
Test added.

### CODE REVIEW

**Comment by robobun** [comment]
@cirospaciari, your commit 4e0ffb5eccab585cfb0658f01ac6ce640775da2e is building: [`#5087`](https://buildkite.com/bun/bun/builds/5087)

File: src/js/node/tls.ts [hunk 1]
```diff
@@ -330,6 +330,7 @@ const TLSSocket = (function (InternalTLSSocket) {
     #socket;
     #checkServerIdentity;
     #session;
+    alpnProtocol = null;
```
Comment (Jarred-Sumner) [inline src/js/node/tls.ts:333]: Does node default to null here? If you want to make sure the property is declared, it is better to use undefined I think.

Comment (Jarred-Sumner) [reply src/js/node/tls.ts:333]: Answer: yes, `null`:

https://github.com/nodejs/node/blob/2eff28fb7a93d3f672f80b582f664a7c701569fb/lib/_tls_wrap.js#L533-L546

File: src/js/node/net.ts [hunk 2]
```diff
@@ -705,7 +718,6 @@ const Socket = (function (InternalSocket) {
         this.#final_callback = callback;
       } else {
         // emit FIN not allowing half open
-        this[bunSocketInternal] = null;
```
Comment (cirospaciari) [inline src/js/node/net.ts:709]: lets only detach on _destroy so we can get certs and more info after this

File: src/js/node/net.ts [hunk 3]
```diff
@@ -684,14 +688,23 @@ const Socket = (function (InternalSocket) {
...
+
+      if (this.writableFinished) {
+        // closed we can detach the socket
         this[bunSocketInternal] = null;
-        // we still have a socket, call end before destroy
-        process.nextTick(endNT, socket, callback, err);
-        return;
+      } else {
+        // lets wait for the finish event before detaching the socket
+        this.once("finish", detachAfterFinish);
```
Comment (cirospaciari) [inline src/js/node/net.ts:709]: lets make sure to only detach after finish and to close both sides so "close" will always be emitted this same code is used on http2, the test added uses the "close" event if you comment the alpnProtocol part will hang in older versions.

File: src/js/node/net.ts [hunk 4]
```diff
@@ -329,7 +331,8 @@ const Socket = (function (InternalSocket) {
         } else {
           self.authorized = true;
         }
-        self.server.emit("secureConnection", self);
+        server[bunSocketServerOptions]?.connectionListener?.$call(server, self);
```
Comment (Jarred-Sumner) [inline src/js/node/net.ts]: If connectionListener is undefined, this will error

is that intentional?

Comment (cirospaciari) [reply src/js/node/net.ts]: was not intentional fixed in https://github.com/oven-sh/bun/pull/14695/commits/bf3f328a659189400894deca043032a91a66f71f

**Pull request merged**